src/batch/src/executor/join/lookup_join.rs

```diff
@@ -257,95 +266,147 @@ impl<P: 'static + ProbeSideSourceBuilder> LookupJoinExecutor<P> {
-            let data_chunk = data_chunk?;
+        while let Some(build_chunk) = build_side_stream.next().await {
+            let build_chunk = build_chunk?.compact()?;
 
-            for row_idx in 0..data_chunk.capacity() {
-                let (cur_row, is_visible) = data_chunk.row_at(row_idx)?;
-                if !is_visible {
-                    continue;
-                }
+            // Group rows with the same key datums together
+            let groups = build_chunk.rows().into_group_map_by(|row| {
+                self.build_side_key_idxs
+                    .iter()
+                    .map(|&idx| row.value_at(idx).to_owned_datum().unwrap())
+                    .collect_vec()
+            });
 
-                let probe_child = self.probe_side_source.build_source(&cur_row).await?;
-                let mut probe_side_stream = probe_child.execute();
-
-                let mut chunk_added = false;
-
-                // The handling of the different join types and which operation they should perform
-                // under each condition is very complex. The gist of it is:
-                //
-                // Inner Join: Always do_inner_join until probe_side_chunk is none.
-                //
-                // Left Outer Join: Always do_inner_join until probe_side_chunk is None. If this
-                //      occurs and no chunk has been added yet, null-pad the current row and add it.
-                //
-                // Left Semi Join: Always do_left_semi_anti_join until probe_side_chunk is None or
-                //      a chunk has been added, whichever is first.
-                //
-                // Left Anti Join: Always do_left_semi_anti_join until probe_side_chunk is None or
-                //      current row has a match on probe side, whichever is first. If the former
-                //      occurs and no match has been found yet, then add current row.
-                // TODO: Simplify the logic of handling the different join types
-                loop {
-                    let probe_side_chunk = probe_side_stream.next().await;
-                    let chunk_is_some = probe_side_chunk.is_some();
-
-                    let join_result = if let Some(chunk) = probe_side_chunk {
-                        let chunk = chunk?;
-
-                        match self.join_type {
-                            JoinType::Inner | JoinType::LeftOuter => {
-                                self.do_inner_join(&cur_row, chunk)
-                            }
-                            JoinType::LeftSemi | JoinType::LeftAnti if !chunk_added => {
-                                self.do_left_semi_anti_join(&cur_row, chunk)
+            let mut row_keys = vec![];
+            let mut all_row_refs = vec![];
+
+            groups.into_iter().for_each(|(row_key, row_refs)| {
+                row_keys.push(row_key);
+                all_row_refs.push(row_refs);
+            });
```

Comment:
Is there any way to get the ordering of the `row_keys` and `row_refs` to be consistent or deterministic? Their inconsistency makes the tests fail sometimes.

Reply:
Alternatively, an option can be added to the tester to make it ignore the order of rows in data chunks when comparing them.